Frontend Overhaul blog post

xoscar opened the pull request:
Hello everyone,

Our team wrote a blog post about the frontend changes a couple of weeks ago, and there were some ideas on possibly having the post added to the open telemetry website as well. We have made some changes to make it vendor-free and better suited for the open telemetry website.

Let me know if you have any questions or concerns!

---

Preview: https://deploy-preview-1877--opentelemetry.netlify.app/blog/2022/frontend-overhaul/

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+
+![Open Telemetry Demo System Diagram](system-diagram.png)
+
+The front end was constituted by a Golang SSR application which sent the
```
Rewording:
“Prior to re-architecting, the front end consisted of a Golang SSR app, which sent complete HTML to the browser for display. Every request and call redirected to the server so new information was shown.”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+
+## Frontend Application Architecture Overhaul
+
+We worked on an initial proposal that included the following bullet points:
```
Rewording:
“… that included the following:”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+- Deployment & Distribution
+- Testing (E2E, unit test)
+
+This proposal was presented to the OpenTelemetry Demo SIG duringone of the
```
“duringone” should be “during one”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+weekly Monday meetings and we were given the green light to move ahead. As part
+of the changes, we decided to use Next.js to not only work as the primary
+frontend application but also to work as an aggregation layer between the
+frontend and the GRPC backend services.
```
Keep “front end” or “frontend” consistent throughout. Same with “backend”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+
+## OpenTelemetry Instrumentation
+
+The next big thing we worked on was having a way to instrument both sides of the
```
Rewording:
“… was a way to instrument both sides of the Next.js app. To do this we had to connect the app twice to the same collector used by all the microservices.”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+twice to the same open telemetry collector that was being used by all of the
+other microservices.
+
+For the backend side, a simple solution was designed that involved using the
```
Rewording:
“A simple backend solution was designed using the …”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+The basic instrumentation includes auto instrumentation for most of the commonly
+used
+[libraries and tools for Node.js](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node).
+But as part of having a better example for users, a manual instrumentation piece
```
Rewording:
“As part of providing a better example for users, a manual instrumentation in the form of route middleware was added. This would catch the …”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+request and create a span based on it, including the context propagation. The
+[implementation can be found here](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frontend/utils/telemetry/InstrumentationMiddleware.ts).
+
+For the frontend side, it became a little bit tricky, as the
```
Rewording:
“The front end was a little trickier, as the rendering is server-side. We had to make sure to …”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+comes from the server side so, in this case, we had to make sure to load the
+tracer from the browser side when the Javascript code is executed.
+
+After adding validations to check for the browser side, we then can proceed to
```
Rewording:
“…check the browser side, we then loaded the custom front end tracing module, which included creating the web tracer …”

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+After adding validations to check for the browser side, we then can proceed to
+load the custom front-end tracing module which includes the creation of the
+[web tracer provider and the automatic web instrumentations](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frontend/utils/telemetry/FrontendTracer.ts).
+The automatic web instrumentations capture the most common user interactions
```
Rewording:
“The automatic front end instrumentation captures the most common user actions such as clicks, fetch requests, and page loads. In order to allow the browser side to interact with the collector, a config change is needed: enable incoming CORS requests from the web app. The resulting collector receivers looks something like this:”

melbgirl reviewed (changes requested):
Mostly language changes/rewording to improve flow and readability.

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+---
+title: Frontend Overhaul of the OpenTelemetry Demo (Go to Next.js)
+linkTitle: Frontend Overhaul
```
```suggestion
linkTitle: OTel Demo Frontend Overhaul
```

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+linkTitle: Frontend Overhaul
+date: 2022-10-18
+author: "[Oscar Reyes](https://github.com/xoscar) (Tracetest)"
+draft: true
```
remove

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+---
+
+One of the OpenTelemetry Project's many Special Interest Groups (SIG) is the
+OpenTelemetry Community Demo SIG which gives support to a set of instrumented
```
```suggestion
[OpenTelemetry Community Demo SIG](https://github.com/open-telemetry/opentelemetry-demo) which gives support to a set of instrumented
```

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+client (CSR), as well as improving the overall style, theme, and user
+experience.
+
+A fun aspect of the work was the request to move the store from a ‘normal’ store
```
```suggestion
A fun aspect of the work was the request to move the store from a "normal" store
```

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+contributions could be the most impactful. We started looking at
+[the issue created by Austin Parker](https://github.com/open-telemetry/opentelemetry-demo/issues/39)
+referencing a complete front-end overhaul that would involve moving the
+application away from Go (SSR) to an architecture that included a browser-side
```
what does SSR, CSR stand for? don't expect people to know that, introduce those abbreviation or link to a definition eg wikipedia

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,182 @@
+from the web app. Having the collector receivers configuration looking similar
+to this:
+
+![OpenTelemetry Collector CORS Config](collector-cors-config.png)
```
don't use a picture, where no picture is needed, some people can not see and some people might want to copy out the content of that picture easily.

svrnm reviewed (changes requested):
thanks @xoscar!

@open-telemetry/demo-approvers please give this a look as well

xoscar commented:
Hello @melbgirl & @svrnm thanks for your review! I really appreciate it 😄 

I have addressed all of your comments, let me know if there is anything else I can do. Cheers!

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+---
+title: Front end Overhaul of the OpenTelemetry Demo (Go to Next.js)
```
"Frontend", "Front-end" or "Front end"?

https://en.wikipedia.org/wiki/Front-end_web_development

svrnm replied:
I think I saw "back-end" used a lot across our docs, so it would be "Front-end" then

melbgirl replied:
Sold!

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+The demo’s stack includes multiple microservices in different languages,
+covering each of the following OTel-supported languages:
+
+1. Go
+2. PHP
+3. C++
+4. .NET
+5. Ruby
+6. Python
+7. Node.js
+8. Rust
+9. Elixir
```
Alphabetic ordering?

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+only help the community, but also provide a great example to test
+and showcase our product with.
+
+The first thing we did was to get in contact with Carter Socha, the organizer of
```
```suggestion
The first thing we did was to get in contact with [Carter Socha](link to GitHub profile), the organizer of
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+9. Elixir
+
+Every microservice has a specific goal and can communicate with others by using
+a global GRPC definition. Persistent information is saved into a Postgres
```
https://en.wikipedia.org/wiki/GRPC

```suggestion
a global gRPC definition. Persistent information is saved into a Postgres
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+a global GRPC definition. Persistent information is saved into a Postgres
+database and there are outbound services that connect with third-party services
+to trigger events (such as confirmation emails). All of the microservices,
+including the front end, are connected to the same Open Telemetry collector
```
```suggestion
including the front end, are connected to the same OpenTelemetry collector
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+database and there are outbound services that connect with third-party services
+to trigger events (such as confirmation emails). All of the microservices,
+including the front end, are connected to the same Open Telemetry collector
+instance, which uses Jaeger as one of the data stores for the traces and spans.
```
It sounds like traces and spans are different things, consider some rewording.

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+
+![Open Telemetry Demo System Diagram](system-diagram.png)
+
+Prior to re-architecting, the front end consisted of a Golang Server Side Rendered app,
```
Why uppercase "Server Side Rendered"?

melbgirl replied:
Since it's been explained earlier in the post, the first time the acronym was used (I think it's line 33), you can use "SSR" here onwards.

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+
+![New Front end Data Flow](data-flow.png)
+
+As you can see in the Diagram, the application has two major connectivity
```
```suggestion
As you can see in the diagram, the application has two major connectivity
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+The front end was a little trickier, as the
+[initial rendering is server-side](https://nextjs.org/learn/foundations/how-nextjs-works/rendering).
+We had to make sure to load the tracer from the browser side when
+the Javascript code is executed.
```
```suggestion
the JavaScript code is executed.
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+fetch requests, and page loads. In order to allow the browser side to interact with the collector,
+a config change is needed: enable incoming CORS requests from the web app.
+
+Once the setup is complete, by loading the application from docker and
```
```suggestion
Once the setup is complete, by loading the application from Docker and
```

reyang commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+enjoyed the process of making a contribution to the OpenTelemetry project and
+are actively looking for more opportunities to contribute!
+
+_Oscar Reyes and Olly Babiak also are working on Tracetest, an open source tool
```
Consider link to GitHub profiles.

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,175 @@
+contributions could be the most impactful. We started looking at
+[the issue created by Austin Parker](https://github.com/open-telemetry/opentelemetry-demo/issues/39)
+referencing a complete front-end overhaul that would involve moving the
+application away from Go (Server Side Render) to an architecture that included a browser-side
```
If you want to use the acronym SSR you totally can. Best practice is to explain the acronym the first time you use it in a post, and then use the acronym after that. 
eg:
" ... away from Go Server-Side Renderer (SSR) to an architecture that included a browser-side client (Client-Side Renderer or CSR), as well as ..."

xoscar commented:
Hello @reyang thanks for your feedback, I think I got it all fixed now. Let me know if there is something else I can do! 😄

reyang commented:
> Hello @reyang thanks for your feedback, I think I got it all fixed now. Let me know if there is something else I can do! 😄

Thanks @xoscar! Consider fixing the diagram, too.
In addition, the diagram has very small fonts (typefaces) which is hard to read, consider making it bigger.

![image](https://user-images.githubusercontent.com/17327289/197040762-3b518c12-1ef0-4b5e-bca7-dace81d68e00.png)

reyang reviewed (approved):
LGTM with some minor suggestions.

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -1,88 +1,91 @@
-client (Client Side Render), as well as improving the overall style, theme, and user
-experience.
+application away from Go server-side render (SSR) to an architecture that
+included a browser-side client (client-side render), as well as improving the
```
Should probably add the acronym for client-side render, ie:  "...client (client-side render or CSR), as ..."

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -1,88 +1,91 @@
+![OpenTelemetry Demo Old Front-end](old-design.png)
 
 The first task at hand was to bring the demo to the modern age by updating the
 design, color schemes, and user experience. Olly Babiak walked into the fray to
```
Would be good to provide a link to Olly Babiak's GitHub profile (or Twitter if GH not available).

melbgirl commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -108,25 +111,25 @@ We worked on an initial proposal that included the following:
 
 This proposal was presented to the OpenTelemetry Demo SIG during one of the
```
The SIG has been referred to as "demo SIG" throughout, suggest applying the lowercase 'd' here for consistency. Not a blocking item!

xoscar commented:
Hey @svrnm & @melbgirl thanks for your comments, I think I have fixed all of them 😄

svrnm commented:
TY @xoscar , minor NIT: the image `system-diagram.png` has some visual icons on the top right corner that make PHP invisible, can you fix that please?

svrnm commented:
@open-telemetry/blog-approvers , please review & approve

cartermp commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,180 @@
+aggregation layer and the other from the aggregation layer to the back-end
+services (gRPC).
+
+## OpenTelemetry Instrumentation
```
One thing that would be really great to have here, and I won't block on this but I still think it'd be helpful, would be an elaboration on how instrumentation is plugging into next.js. Specifically, I see that there's a next.js middleware implementation that creates spans. That's cool! But why is it there, and what does it do? How does it get used? I think that would be _very_ interesting to people who are also getting into tracing with frontends.

cartermp replied:
Yep, that sounds neat! Maybe that could also make more sense as a readme or something in the demo repo itself. It's just something I got curious about since the post had me digging through code and it seemed cool but wasn't clear on the approach.

svrnm commented on content/en/blog/2022/frontend-overhaul/index.md:
````diff
@@ -0,0 +1,265 @@
+including the front-end, are connected to the same OpenTelemetry collector
+instance, which uses Jaeger as one of the data stores for the tracing data.
+
+```mermaid
````
while github has a mermaid support, we do not have it in hugo, so you still need to put the picture here.

xoscar replied:
Ups, sorry about that, I just fixed it with my latest commit. Thanks for the heads up! 😄

mic-max commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+application, no matter what programming language, platform or OS it uses. The
+web app also shows different instrumentation techniques: automatic and manual,
+metrics, and baggage. All while following the standards and conventions
+proscribed in the official OTel documentation. More about the specific
```
typo ```suggestion
prescribed in the official OTel documentation. More about the specific
```

cartermp commented:
@open-telemetry/blog-approvers PTAL!

chalin commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+A simple back-end solution was designed using the
+[official gRPC exporter](https://www.npmjs.com/package/@opentelemetry/exporter-trace-otlp-grpc)
+in combination with the
+[Nodejs SDK](https://www.npmjs.com/package/@opentelemetry/sdk-node).
```
Nit:
```suggestion
[Node.js SDK](https://www.npmjs.com/package/@opentelemetry/sdk-node).
```

chalin commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+The demo’s stack includes multiple microservices in different languages,
+covering each of the following OTel-supported languages:
+
+- .NET
+- C++
+- Elixir
+- Go
+- Node.js
+- PHP
+- Python
+- Ruby
+- Rust
```
Would it make sense to make these links into the corresponding language SIG's landing page in the website?

xoscar replied:
Hello @chalin do you know where I can find the links to the SIG's landing page?

chalin replied:
I was referring to the listing of languages under https://opentelemetry.io/docs/instrumentation/

xoscar replied:
Done 😄

chalin reviewed (commented):
See inline for question and minor copy-edit.

chalin commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+The demo’s stack includes multiple microservices in different languages,
+covering each of the following OTel-supported languages:
+
+- [.NET](https://opentelemetry.io/docs/instrumentation/net/)
+- [C++](https://opentelemetry.io/docs/instrumentation/cpp/)
+- [Erlang/Elixir](https://opentelemetry.io/docs/instrumentation/erlang/)
+- [Go](https://opentelemetry.io/docs/instrumentation/go/)
+- [Node.js](https://opentelemetry.io/docs/instrumentation/js/)
+- [PHP](https://opentelemetry.io/docs/instrumentation/php/)
+- [Python](https://opentelemetry.io/docs/instrumentation/python/)
+- [Ruby](https://opentelemetry.io/docs/instrumentation/ruby/)
+- [Rust](https://opentelemetry.io/docs/instrumentation/rust/)
```
Changes:

- Use local paths not external URLs here
- Alphabetical order of languge names: put .NET after C++ (as if it were named dotnet).

```suggestion
- [C++](/docs/instrumentation/cpp/)
- [.NET](/docs/instrumentation/net/)
- [Erlang/Elixir](/docs/instrumentation/erlang/)
- [Go](/docs/instrumentation/go/)
- [Node.js](/docs/instrumentation/js/)
- [PHP](/docs/instrumentation/php/)
- [Python](/docs/instrumentation/python/)
- [Ruby](/docs/instrumentation/ruby/)
- [Rust](/docs/instrumentation/rust/)
```

chalin reviewed (commented):
See inline comment for a suggested change that will fix the build.

xoscar commented:
@chalin I don't want to be pushy but, is there something else I can do to have this one merged? 😄

chalin commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+---
+title: Front-end Overhaul of the OpenTelemetry Demo (Go to Next.js)
+linkTitle: OTel Demo Front-end Overhaul
```
```suggestion
linkTitle: Front-end Overhaul Demo
```

chalin reviewed (approved):
One suggested (and optional) short-title tweak. Otherwise LGTM ✨

chalin commented:
@xoscar - @svrnm and @cartermp will time the merge (and tweak the publish date?), once ready. I've given my final approval, thanks.

chalin commented on content/en/blog/2022/frontend-overhaul/index.md:
```diff
@@ -0,0 +1,181 @@
+---
+title: Front-end Overhaul of the OpenTelemetry Demo (Go to Next.js)
+linkTitle: Front-end Overhaul Demo
+date: 2022-10-18
```
```suggestion
date: 2022-11-07
draft: true
```

chalin replied:
Draft status as agreed during Comms meeting.

cartermp commented:
Thanks @xoscar! We will un-draft this next Monday (discussed in Comms SIG) and have it get social media boostage.

svrnm commented:
@xoscar see #2020
